arch: Add SRAM based context for ISR

Martinhoff-maker opened the pull request:
Edit after rework:

The goal of this PR is to add the symbol CONFIG_SRAM_SW_ISR_TABLE which allows placing the Software ISR table in RAM. This was previously possible with CONFIG_DYNAMIC_INTERRUPTS, but that option allows write access to the ISR table and includes functions to insert interrupts dynamically, which we don't want.

Additionally, it adds a test to ensure that it's possible to relocate everything needed to have a full RAM-based context for an ISR. It relocates the entire vector table and necessary files (isr_wrapper, etc.) to RAM in order to have a fully RAM-based context when an interrupt is triggered. This eliminates latency (from accessing flash) when an interrupt is triggered in a real-time application.

danieldegrasse commented:
Not inherently opposed to the idea here, but this seems like a slippery slope. If you are trying to avoid _all_ flash access during an ISR, this change won't be sufficient (data like device pointers may still be in flash, even if you move relevant ISR handers). As it stands, I'm kind of unclear what the advantage of this change is over turning on `CONFIG_DYNAMIC_INTERRUPTS` and directly relocating architecture files from your application CMakeLists.

Is there something I'm missing here, or is this Kconfig currently mostly for convenience (IE something you could accomplish within your application with a bit more work)?

Martinhoff-maker commented:
Thanks for the review @danieldegrasse. I'm currently aware about the device pointer. I'm trying to be as generic as possible and in the case that you relocate an ISR that is not related to a driver, you don't need that part. The whole point is to have a starting point for people who trying to have a real time constraint on an ISR. That's why I say, even in the Kconfig helper, that the relocation of the ISR handler and all of his dependency (and this includes device struct for example) must be done by the user.

CONFIG_DYNAMIC_INTERRUPT include a lot of files that we don't need for this feature, that's why I prefer to split it into 2 parts.

However, It needs to be well documented I think in order to not be a slippery option (explain what user have to take care about in case of a driver, and also how to identify the dependency of the ISR handler (with Puncover for example)).

nordicjm commented on arch/Kconfig:
```diff
@@ -589,6 +589,22 @@ config SRAM_SW_ISR_TABLE
 	  The option specifies that the software interrupts vector table will be
 	  placed inside SRAM instead of the flash.
 
+
```
```suggestion
```

nordicjm commented on arch/arm/core/cortex_m/CMakeLists.txt:
```diff
@@ -71,3 +71,8 @@ zephyr_linker_sources_ifdef(CONFIG_SRAM_VECTOR_TABLE
 endif()
+
+if (CONFIG_SRAM_ISR_EXEC_CONTEXT)
+zephyr_code_relocate(FILES ${ZEPHYR_BASE}/arch/arm/core/cortex_m/isr_wrapper.c LOCATION RAM)
+zephyr_code_relocate(FILES ${ZEPHYR_BASE}/arch/arm/core/cortex_m/irq_init.c LOCATION RAM)
```
missing indent (2 spaces for cmake)

nordicjm commented on arch/arm/core/cortex_m/CMakeLists.txt:
```diff
@@ -71,3 +71,8 @@ zephyr_linker_sources_ifdef(CONFIG_SRAM_VECTOR_TABLE
 )
 endif()
+
+if (CONFIG_SRAM_ISR_EXEC_CONTEXT)
```
no space between if and bracket in cmake

Martinhoff-maker replied:
done

danieldegrasse commented:
> Thanks for the review @danieldegrasse. I'm currently aware about the device pointer. I'm trying to be as generic as possible and in the case that you relocate an ISR that is not related to a driver, you don't need that part. The whole point is to have a starting point for people who trying to have a real time constraint on an ISR. That's why I say, even in the Kconfig helper, that the relocation of the ISR handler and all of his dependency (and this includes device struct for example) must be done by the user.
> 
> CONFIG_DYNAMIC_INTERRUPT include a lot of files that we don't need for this feature, that's why I prefer to split it into 2 parts.
> 
> However, It needs to be well documented I think in order to not be a slippery option (explain what user have to take care about in case of a driver, and also how to identify the dependency of the ISR handler (with Puncover for example)).

That makes sense regarding `CONFIG_DYNAMIC_INTERRUPT`. Could we do something like the following?
- make `CONFIG_SRAM_SW_ISR_TABLE` a user-selectable Kconfig, that `CONFIG_DYNAMIC_INTERRUPT` enables
- Rely on the user to enable `CONFIG_SRAM_VECTOR_TABLE`, `CONFIG_SRAM_SW_ISR_TABLE`, and use `zephyr_code_relocate` to move files to SRAM

I think my core concern is how many caveats we have to place on the Kconfig `SRAM_ISR_EXEC_CONTEXT`- I worry that it will lead users to turn this on without fully understanding what is and isn't relocated, whereas if they can set `CONFIG_SRAM_SW_ISR_TABLE` they have a good idea of what Zephyr has located in RAM, and what they need to relocate themselves.

Martinhoff-maker commented:
I totally agree with what you are saying. I just fear that somebody who wants to have an ISR RAM context will have to go through all the works that I have already done (for cortex-M). Maybe the only expression of this work might be a good documentation but where ? maybe in the code relocation ? or in the interrupt documentation ?

danieldegrasse commented:
> I totally agree with what you are saying. I just fear that somebody who wants to have an ISR RAM context will have to go through all the works that I have already done (for cortex-M). Maybe the only expression of this work might be a good documentation but where ? maybe in the code relocation ? or in the interrupt documentation ?

What about adding a sample demonstrating how to perform the relocation? That way users have somewhere to start and some documentation on how to continue from there

wearyzen commented on arch/Kconfig:
```diff
@@ -462,6 +462,7 @@ config ISR_TABLES_LOCAL_DECLARATION
 
 config DYNAMIC_INTERRUPTS
 	bool "Installation of IRQs at runtime"
+	select SRAM_SW_ISR_TABLE
```
should be "This commit introduce **SRAM_SW_ISR_TABLE** hidden option which is" in the commit message

Martinhoff-maker commented:
Agree with all of you, I will replace my last with an example.

Martinhoff-maker commented on tests/application_development/ram_context_for_isr/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+  zephyr_code_relocate(FILES ${ZEPHYR_BASE}/arch/arm/core/cortex_m/exc_exit.c LOCATION RAM)
+endif()
+
+zephyr_code_relocate(FILES ${CMAKE_CURRENT_SOURCE_DIR}/src/main.c FILTER ".test_irq_callback" LOCATION RAM )
```
done

Martinhoff-maker commented on tests/application_development/ram_context_for_isr/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+endif()
+
+zephyr_code_relocate(FILES ${CMAKE_CURRENT_SOURCE_DIR}/src/main.c FILTER ".test_irq_callback" LOCATION RAM )
+zephyr_code_relocate(FILES ${CMAKE_CURRENT_SOURCE_DIR}/src/fake_driver.c FILTER ".fake_driver_isr" LOCATION RAM )
```
done

Martinhoff-maker commented on tests/application_development/ram_context_for_isr/README.rst:
```diff
@@ -0,0 +1,94 @@
+A test that verifies the ISR vector table relocation and
+full IRQ execution under RAM context.
+
+
```
done

Martinhoff-maker commented on tests/application_development/ram_context_for_isr/testcase.yaml:
```diff
@@ -0,0 +1,19 @@
+    arch_allow: arm
+    filter: CONFIG_CPU_CORTEX_M_HAS_VTOR
+    extra_configs:
+      - CONFIG_CODE_DATA_RELOCATION_SRAM=y
+      # In case you want to add this config, place all the possible called code in RAM
+      - CONFIG_TRACING_ISR=n
+      - CONFIG_STACK_SENTINEL=n
+      - CONFIG_PM=n
```
I don't think so since it's very specific to the ARM architecture. What we need to exclude depends on each architecture and how it handles ISR handlers.

Martinhoff-maker replied:
for the moment yes but it is to prepare if other arch want to be added at this test.
Moreover, for CONFIG_CODE_DATA_RELOCATION_SRAM, it needs to be only activated if CPU_CORTEX_M then it needs the filter.

Martinhoff-maker commented on tests/application_development/ram_context_for_isr/README.rst:
I'm trying to find a good place for this then, maybe in zephyr/doc/hardware/arch just after the "Code relocation" section ? I don't find any doc regarding linker/Cortex-M

Martinhoff-maker commented on doc/kernel/services/interrupts.rst:
```diff
@@ -197,6 +197,10 @@ generate exceptions that need to be handled synchronously (e.g. kernel panic).
 
+.. tip::
+    To mitigate flash access latency, consider relocating ISRs and all related
+    symbols to RAM. This eliminates flash access delays.
```
done

Martinhoff-maker commented on doc/kernel/services/interrupts.rst:
```diff
@@ -374,6 +378,16 @@ Cortex-M architecture variant via the macro
+:kconfig:option:`CONFIG_SRAM_SW_ISR_TABLE` will help you have interrupt vectors
+in RAM. :kconfig:option:`CONFIG_DEVICE_MUTABLE` will help you put the device
+structure in RAM and :kconfig:option:`CONFIG_CODE_DATA_RELOCATION` will help you relocate
+the ISR and all the related symbols to RAM.
```
done

JarmouniA reviewed (dismissed):
The PR description needs an update for accuracy.

Martinhoff-maker commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -83,3 +83,6 @@ Modules
 
 Architectures
 *************
+
+* The :kconfig:option:`CONFIG_DYNAMIC_INTERRUPTS` option has a new dependency on
+  :kconfig:option:`CONFIG_SRAM_SW_ISR_TABLE`.
```
There is nothing to do for the user, it is just an intermediate kconfig option, maybe it's just not needed in the migration guide ?

fabiobaltieri replied:
yup, drop it

Martinhoff-maker committed:
tests: application_development: add new test for ram-based isr

This commit introduces the new test ram_context_for_isr that shows
how to configure an application (prj.conf, CMakeLists.txt, etc.) in
order to enable full ISR execution with a RAM context.

It resolves the issue of flash latency affecting real-time constraints
during ISR execution.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
arch: split dynamic interrupt symbol

This commit introduces the SRAM_SW_ISR_TABLE option which is selected by
DYNAMIC_INTERRUPT. It allows splitting the DYNAMIC_INTERRUPT option into
two parts:

 - One for the relocation of the ISR vector table in RAM
 - One for the inclusion of functions needed to install ISRs dynamically

The goal is to later only select the relocation of the ISR vector table in
RAM and not all the associated functions from the dynamic interrupt
mechanism.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker commented:
gentle ping to @JarmouniA and @wearyzen (normally the last time 👍 )

mathieuchopstm commented:
Broken on platforms with only 2 priority bits. Fix PR :point_right: #93907
